projects/compilation.vala

```diff
@@ -0,0 +1,209 @@
+    private HashSet<string> _metadata_dirs;
+    private HashSet<string> _gresources_dirs;
+    private HashSet<string> _defines;
+
+    // compiler flags
+
+    public bool experimental { get; private set; }
+    public bool experimental_non_null { get; private set; }
+    public Vala.Profile profile { get; private set; }
+    public bool abi_stability { get; private set; }
+
+    // code context
+    /**
+     * Whether the code context needs to be (re-)compiled.
+     */
+    public bool needs_compile { get; private set; default = true; }
+
+    /**
+     * Whether the code context has been updated. This could happen
+     * because we added a new source file, or we changed a source file,
+     * for example.
+     */
+    public bool dirty { get; private set; default = true; }
+
+    private Vala.CodeContext? _ctx;
+
+    public Vala.CodeContext code_context {
+        get {
+            if (dirty) {
+                debug ("dirty context, rebuilding");
+
+                if (_ctx != null) {
+                    // stupid workaround for memory leaks in Vala 0.38
+                    workaround_038 (_ctx, _sources.values);
+                }
+                // generate a new code context 
+                _ctx = new Vala.CodeContext () { keep_going = true };
+                Vala.CodeContext.push (_ctx);
+                dirty = false;
+
+                // string version = Config.libvala_version;
+                _ctx.report = new Reporter ();
+
+                // compiler flags and defines
+                _ctx.profile = profile;
+                switch (profile) {
+                    case Vala.Profile.POSIX:
+                        _ctx.add_define ("POSIX");
+                        break;
+                    default: // case Vala.Profile.GOBJECT:
+                        _ctx.add_define ("GOBJECT");
+                        break;
+                }
+
+                _ctx.experimental = experimental;
+                _ctx.experimental_non_null = experimental_non_null;
+                _ctx.abi_stability = abi_stability;
+
+                _ctx.vapi_directories = _vapi_dirs.to_array ();
+                _ctx.gir_directories = _gir_dirs.to_array ();
+                _ctx.metadata_directories = _metadata_dirs.to_array ();
+                _ctx.gresources_directories = _gresources_dirs.to_array ();
+
+                // packages
+                _ctx.add_external_package ("glib-2.0");
+                _ctx.add_external_package ("gobject-2.0");
+                foreach (var package in _packages)
+                    _ctx.add_external_package (package);
+
+                foreach (TextDocument doc in _sources.values) {
+                    doc.file.context = _ctx;
+                    _ctx.add_source_file (doc.file);
+                    // clear all using directives (to avoid "T ambiguous with T" errors)
+                    doc.file.current_using_directives = new Vala.ArrayList<Vala.UsingDirective> ();
```

Comment:
`doc.file.current_using_directives.clear()`

Reply:
Yeah, this is from a while ago. I don't really want to explore changing it in this MR since some of these odd pieces of code came about when I was trying to aggressively limit memory consumption.